usersync/syncersbuilder.go

```diff
@@ -98,28 +101,88 @@ func chooseSyncerConfig(biddersSyncerConfig []namedSyncerConfig) (namedSyncerCon
 		return biddersSyncerConfig[0], nil
 	}
 
-	var bidderNames []string
-	var bidderNamesWithEndpoints []string
-	var syncerConfig namedSyncerConfig
+	var (
+		bidderNames         []string
+		nonAliasBidderNames []string
+		aliasBidderNames    []string
+		syncerConfig        namedSyncerConfig
+		parentBidderNameMap = make(map[string]struct{})
+	)
+
 	for _, bidder := range biddersSyncerConfig {
 		bidderNames = append(bidderNames, bidder.name)
 		if bidder.cfg.IFrame != nil || bidder.cfg.Redirect != nil {
-			bidderNamesWithEndpoints = append(bidderNamesWithEndpoints, bidder.name)
+			if len(bidder.bidderInfo.AliasOf) > 0 {
+				aliasBidderNames = append(aliasBidderNames, bidder.name)
+				parentBidderNameMap[bidder.bidderInfo.AliasOf] = struct{}{}
+			} else {
+				nonAliasBidderNames = append(nonAliasBidderNames, bidder.name)
+			}
 			syncerConfig = bidder
 		}
 	}
 
-	if len(bidderNamesWithEndpoints) == 0 {
+	// check if bidders have same syncer key but no endpoints
+	if len(nonAliasBidderNames)+len(aliasBidderNames) == 0 {
 		sort.Strings(bidderNames)
 		bidders := strings.Join(bidderNames, ", ")
 		return namedSyncerConfig{}, fmt.Errorf("bidders %s share the same syncer key, but none define endpoints (iframe and/or redirect)", bidders)
 	}
 
-	if len(bidderNamesWithEndpoints) > 1 {
-		sort.Strings(bidderNamesWithEndpoints)
-		bidders := strings.Join(bidderNamesWithEndpoints, ", ")
+	// check if non-alias bidders have same syncer key
+	if len(nonAliasBidderNames) > 1 {
+		sort.Strings(nonAliasBidderNames)
+		bidders := strings.Join(nonAliasBidderNames, ", ")
 		return namedSyncerConfig{}, fmt.Errorf("bidders %s define endpoints (iframe and/or redirect) for the same syncer key, but only one bidder is permitted to define endpoints", bidders)
 	}
 
+	// check if alias bidders of different parent have same syncer key
+	if len(parentBidderNameMap) > 1 {
+		sort.Strings(aliasBidderNames)
+		return namedSyncerConfig{}, fmt.Errorf("alias bidders %s of different parents defines endpoints (iframe and/or redirect) for the same syncer key, but only one bidder is permitted to define endpoints", strings.Join(aliasBidderNames, ", "))
+	}
+
+	// check if aliases of same parent and non-alias bidder have same syncer key
+	if len(parentBidderNameMap) != 0 {
+		if _, ok := parentBidderNameMap[nonAliasBidderNames[0]]; !ok {
+			sort.Strings(aliasBidderNames)
+			return namedSyncerConfig{}, fmt.Errorf("alias bidders %s and non-alias bidder %s defines endpoints (iframe and/or redirect) for the same syncer key, but only one bidder is permitted to define endpoints", strings.Join(aliasBidderNames, ", "), nonAliasBidderNames[0])
+		}
+	}
```

Comment:
Can these two if statements be consolidated at all? If `len(parentBidderNameMap) > 1` is true, then it means that it's != 0. Feels like an opportunity for consolidation.

Reply:
https://github.com/prebid/prebid-server/pull/3082/commits/8a54873dabdf8c209cedadd47588d59e2fb3234f

Did some refactoring. But still  `len(parentBidderNameMap) > 1 ` and `len(parentBidderNameMap) != 0` checks are needed. @AlexBVolcy, could you suggest  change  to consolidate both checks

Reply:
Maybe you could re-write it like this, where you just separate the two cases and remove the overarching `len(parentBidderNameMap) != 0` check instead of nesting them. This is probably more of a nitpick from me, so if you prefer the original implementation, I'm fine with that!

```
if len(parentBidderNameMap) > 1 {
    sort.Strings(aliasBidderNames)
    return namedSyncerConfig{}, fmt.Errorf("either alias bidders %s defines endpoints (iframe and/or redirect) for the same syncer key used by multiple bidders, but only one bidder is permitted to define endpoints", strings.Join(aliasBidderNames, ", "))
}

if _, ok := parentBidderNameMap[nonAliasBidderNames[0]]; !ok && len(parentBidderNameMap) > 0 {
    sort.Strings(aliasBidderNames)
    return namedSyncerConfig{}, fmt.Errorf("either alias bidders %s defines endpoints (iframe and/or redirect) for the same syncer key used by other bidder, but only one bidder is permitted to define endpoints", strings.Join(aliasBidderNames, ", "))
}
```